### PR TITLE
Sharing PathNormalizer Implementations

### PR DESCRIPTION
**Pull request by ladeak**
# Sharing PathNormalizer Implementations

Sharing PathNormalizer Implementations between HttpSys and Kestrel servers

## Description

- Follow up PR on comment: https://github.com/dotnet/aspnetcore/pull/56805#discussion_r1702625899
- Moving PathNormalizer from Shared\HttpSys to Shared\PathNormalizer (any other suggestions welcome), changing the namespace to Microsoft.AspNetCore.Internal
- Sharing this file with HttpSys and Kestrel.Core
- Remaining Kestrel specific code renamed to PathDecoder (open for suggestions)
- Keeping the tests and benchmarks in Kestrel

Linked to #56534

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Looks like you missed including the shared file in the httpsys csproj

**Comment by ladeak** [comment]
Thank you  @BrennanConroy I will follow up on that.

File: src/Servers/Kestrel/Core/src/Internal/Http/PathDecoder.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+using System.Text;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+
+internal static class PathDecoder
+{
+    public static string DecodePath(Span<byte> path, bool pathEncoded, string rawTarget, int queryLength)
```
[the remark this answers is not in the document]
Comment (ladeak) [reply src/Servers/Kestrel/Core/src/Internal/Http/PathDecoder.cs:12]: I had to revert, these are all in-place updates hence they require a writable Span.

[the remark this answers is not in the document]
Comment (gfoidl) [reply src/Servers/Kestrel/Core/src/Internal/Http/PathDecoder.cs:12]: Ah, sorry.

**Review by halter73 — APPROVED** [review]
LGTM, but I'll wait for @BrennanConroy to take a look to merge.

File: src/Servers/Kestrel/Core/test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests.csproj [hunk 2]
```diff
@@ -19,6 +19,7 @@
     <Compile Include="$(SharedSourceRoot)test\Shared.Tests\runtime\Http2\*.cs" LinkBase="Shared\runtime\Http2" />
     <Compile Include="$(SharedSourceRoot)test\Shared.Tests\runtime\Http3\*.cs" LinkBase="Shared\runtime\Http3" />
     <Compile Include="$(SharedSourceRoot)TestResources.cs" LinkBase="Shared" />
+    <Compile Include="$(SharedSourceRoot)test\Shared.Tests\PathNormalizerTests.cs" LinkBase="/" />
```
Comment (BrennanConroy) [inline src/Servers/Kestrel/Core/test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests.csproj:22]: I think this is causing these tests to run again. They should already be running from the Shared.Tests.csproj.

Comment (ladeak) [reply src/Servers/Kestrel/Core/test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests.csproj:22]: Yes, I can remove this line.

Comment (ladeak) [reply src/Servers/Kestrel/Core/test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests.csproj]: Updated. I wonder though how "discoverability" can be improved as the shared files are linked ie. in Kestrel.slnf but changing and finding corresponding unit tests are not that easy as the projects are not linked. Or is there a better slnf to edit.

ie. I did not notice this file exists until the failure in the CI, while I knew about the PathNormaizers.

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**